go/test/endtoend/vtgate/reference/reference_test.go

```diff
@@ -0,0 +1,130 @@
+/*
+Copyright 2022 The Vitess Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package reference
+
+import (
+	"context"
+	"fmt"
+	"testing"
+
+	"github.com/stretchr/testify/require"
+
+	"vitess.io/vitess/go/test/endtoend/utils"
+
+	"vitess.io/vitess/go/mysql"
+	"vitess.io/vitess/go/test/endtoend/cluster"
+)
+
+// TestGlobalReferenceRouting tests that unqualified queries for reference
+// tables go to the right place.
+//
+// Given:
+//   - Unsharded keyspace `uks` and sharded keyspace `sks`.
+//   - Source table `uks.zip_detail` and a reference table `sks.zip_detail`,
+//     initially with the same rows.
+//   - Unsharded table `uks.zip` and sharded table `sks.delivery_failure`.
+//
+// When: we execute `INSERT INTO zip_detail ...`,
+// Then: `zip_detail` should be routed to `uks`.
+//
+// When: we execute `UPDATE zip_detail ...`,
+// Then: `zip_detail` should be routed to `uks`.
+//
+// When: we execute `SELECT ... FROM zip JOIN zip_detail ...`,
+// Then: `zip_detail` should be routed to `uks`.
+//
+// When: we execute `SELECT ... FROM delivery_failure JOIN zip_detail ...`,
+// Then: `zip_detail` should be routed to `sks`.
+//
+// When: we execute `DELETE FROM zip_detail ...`,
+// Then: `zip_detail` should be routed to `uks`.
+func TestReferenceRouting(t *testing.T) {
+	defer cluster.PanicHandler(t)
```

Comment:
we have started using a new pattern for our end2end tests. You can see it in use [here](https://github.com/vitessio/vitess/blob/main/go/test/endtoend/vtgate/queries/aggregation/aggregation_test.go#L52).

The idea is to put the repetitive code we have to write out in any test method used. We set stuff up and return a destructor that knows how to clean out state, so tests don't accidentally step on each other.